test/affinity_tests.cc

```diff
@@ -0,0 +1,323 @@
+	skip_unsupported_configurations();
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "false");
+	raii_affinity_masking mask({0});
+
+	celerity::queue q;
+
+	CHECK(test_utils::log_contains_substring(
+	    detail::log_level::warn, fmt::format("only {} logical cores are available to this process. It is recommended to assign at least", 1)));
+	CHECK_FALSE(test_utils::log_contains_substring(detail::log_level::debug, "Affinity"));
+}
+
+TEST_CASE_METHOD(test_utils::runtime_fixture, "the application thread is actually pinned when pinning is enabled", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "auto");
+
+	const core_set process_mask = {3, 4, 5, 6, 7};
+	if(!have_cores(process_mask)) {
+		SKIP("Skipping test because not all needed cores are available");
+		return;
+	}
+	raii_affinity_masking mask(process_mask);
+
+	{
+		raii_test_runtime rt(1);
+		CHECK(get_current_cores() == core_set{3});
+	}
+
+	// check that the application thread was unpinned on shutdown
+	CHECK(get_current_cores() == process_mask);
+
+	CHECK(test_utils::log_contains_substring(detail::log_level::debug, "Affinity: pinned thread 'application' to core 3"));
+}
+
+TEST_CASE_METHOD(test_utils::runtime_fixture, "when pinning is disabled, no threads are pinned", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "false");
+	const auto initial_core_set = get_current_cores();
+
+	{
+		raii_test_runtime rt(1);
+		CHECK(get_current_cores() == initial_core_set);
+	}
+
+	CHECK(get_current_cores() == initial_core_set);
+	CHECK_FALSE(test_utils::log_contains_substring(detail::log_level::debug, "Affinity"));
+}
+
+TEST_CASE_METHOD(test_utils::runtime_fixture, "rt warns when the application thread changes pinning unexpectedly", "[affinity][runtime]") {
+	skip_unsupported_configurations();
+	test_utils::allow_max_log_level(detail::log_level::warn);
+	env::scoped_test_environment ste("CELERITY_THREAD_PINNING", "auto");
+
+	const core_set process_mask = {3, 4, 5, 6, 7};
+	if(!have_cores(process_mask)) {
+		SKIP("Skipping test because not all needed cores are available");
+		return;
+	}
+	raii_affinity_masking mask(process_mask, false); // no cleanup check, since we aren't clean on purpose in this test
+
+	{
+		raii_test_runtime rt(1);
+		CHECK(get_current_cores() == core_set{3});
+
+		// now change the affinity of the application thread
+		cpu_set_t cpuset;
+		CPU_ZERO(&cpuset);
+		CPU_SET(4, &cpuset);
+		REQUIRE(pthread_setaffinity_np(pthread_self(), sizeof(cpu_set_t), &cpuset) == 0);
+	}
+
+	CHECK(test_utils::log_contains_substring(
+	    detail::log_level::warn, fmt::format("Thread affinity of thread {} was changed unexpectedly, skipping restoration.", pthread_self())));
+}
+
+TEST_CASE("multiple subsequent non-overlapping pinning ranges are handled correctly", "[affinity]") {
```

Comment:
Wait, what is this testing? If the thread pinner goes out of scope, it should restore the previous state, right? Why is the "non-overlapping" part relevant?

Reply:
Kind of, "overlapping lifetime" is clear to me, but "non-overlapping pinning ranges" makes it sound that its about the ranges (which are in fact overlapping).

Reply:
Ah, I'll change the wording.